Sources/SwiftParser/Diagnostics/ParserDiagnosticMessages.swift

```diff
@@ -131,79 +132,184 @@ public struct InvalidIdentifierError: ParserError {
+        }
+
+        switch (parts.last, newPart) {
+        case (.sourceText(let previousContent), .sourceText(let newContent)):
+          parts[parts.count - 1] = .sourceText(previousContent + newContent)
+        default:
+          parts.append(newPart)
+        }
+      } else {
+        parts.append(.node(missingNode))
+      }
+    }
+    let partDescriptions = parts.map({ $0.description })
+    if partDescriptions.count > 1 {
+      return "\(partDescriptions[0..<partDescriptions.count - 1].joined(separator: ", ")) and \(partDescriptions.last!)"
     } else {
-      message += " '\(missingToken.text)'"
+      return "\(partDescriptions.joined(separator: ", "))"
     }
-    if let parent = missingToken.parent, let parentTypeName = parent.nodeTypeNameForDiagnostics() {
+  }
+
+  /// If applicable, returns a string that describes after which node the nodes are expected.
+  private var afterClause: String? {
+    if !missingNodes.first!.is(TokenSyntax.self) {
+      if let missingDecl = missingNodes.first?.as(MissingDeclSyntax.self), let lastModifier = missingDecl.modifiers?.last {
+        return "after '\(lastModifier.name.text)' modifier"
+      } else if let missingDecl = missingNodes.first?.as(MissingDeclSyntax.self), missingDecl.attributes != nil {
+        return "after attribute"
+      } else if let previousToken = missingNodes.first?.previousToken(viewMode: .fixedUp), previousToken.presence == .present {
+        return "after '\(previousToken.text)'"
+      }
+    }
+    return nil
+  }
+
+  /// If applicable, returns a string that describes the node in which the missing nodes are expected.
+  private var parentContextClause: String? {
+    enum ContextType {
+      case toStart
+      case `in`
+      case toEnd
+    }
+    var nodePosition = ContextType.in
+
+    if let missingToken = missingNodes.first?.as(TokenSyntax.self), let commonParent = commonParent {
       switch missingToken.tokenKind {
-      case .leftAngle, .leftBrace, .leftParen, .leftSquareBracket:
-        if parent.children(viewMode: .fixedUp).first?.as(TokenSyntax.self) == missingToken {
-          message += " to start \(parentTypeName)"
+      case .leftBrace, .leftAngle, .leftParen, .leftSquareBracket:
+        if commonParent.children(viewMode: .fixedUp).first?.as(TokenSyntax.self) == missingToken {
+          nodePosition = .toStart
         }
-      case .rightAngle, .rightBrace, .rightParen, .rightSquareBracket:
-        if parent.children(viewMode: .fixedUp).last?.as(TokenSyntax.self) == missingToken {
-          message += " to end \(parentTypeName)"
+      default:
+        break
+      }
+    }
+    if let missingToken = missingNodes.last?.as(TokenSyntax.self), let commonParent = commonParent {
+      switch missingToken.tokenKind {
+      case .rightBrace, .rightAngle, .rightParen, .rightSquareBracket:
+        if commonParent.children(viewMode: .fixedUp).last?.as(TokenSyntax.self) == missingToken {
+          if nodePosition == .toStart {
```

Comment:
`in` is the default, just `if nodePosition != .toStart`?

Reply:
I refactored this logic quite a bit to make it simpler so it doesn’t use `ContextType` anymore.